File: parcels/toast/legacy/toast.js
import "@/bootstrap.scss";

import { Toast } from "bootstrap";

import { create_element } from "@/utils/create_element"; ////


// Create toast container to control stacked toasts
document.body.classList.add("position-relative");
const toast_container = create_element(
  `div.position-fixed.bottom-0.end-0.p-3.d-flex.flex-column.row-gap-3.z-3`,
  { id: "toastContainer", parent: document.body }
);

/* Shows a toast */
export function toast(
  title,
  content,
  {
    animation = true,
    autohide = true,
    delay = 5000,
    dismissible = true,
    style = "primary",
  } = {}
) {
  // Prepare styling
  const header_class = `.text-bg-${style}`;
  let dismiss_class = "";
  if (["danger", "primary", "secondary", "success"].includes(style)) {
    dismiss_class = `.btn-close-white`;
  }

  // Prepare title
  if (title) {
    if (typeof title === "string") {
      title = create_element(`h1.fs-6.text.p-0.m-0`, {}, title);
    }
  }

  // Prepare content
  if (content) {
    if (typeof content === "string") {
      content = create_element(`p.m-0`, {}, content);
    }
  }

  // Handle dismiss button
  if (dismissible) {
    dismissible = create_element(`button.btn-close.ms-auto${dismiss_class}`, {
      type: "button",
    })
      .update_dataset({
        bsDismiss: "toast",
      })
      .update_attrs({ "aria-label": "Close" });
  } else {
    dismissible = "";
  }

  // Create toast element
  const element = create_element(
    `div.toast`,
    { parent: toast_container, role: "alert" },
    create_element(
      `div.toast-header.d-flex.align-items-center${header_class}`,
      {},
      title,
      dismissible
    ),
    create_element(`div.toast-body`, {}, content)
  ).update_attrs({ "aria-live": "assertive", "aria-atomic": "true" });

  // Create Bootstrap Toast
  const toast = new Toast(element, { animation, autohide, delay });

  // Ensure clean-up
  element.addEventListener("hidden.bs.toast", () => {
    toast.dispose();
    element.remove();
  });

  // Show the toast
  toast.show();
}

/*
# EXAMPLES

## Example: Hello World

import toast from "utils/toast";

toast({
  title: "Hello world!",
  content: "The toast function is awesome.",
  delay: 10000,
  style: 'success'
});

*/

File: test/client_code/packages/toast/toast.py
def main(use, *args, **kwargs):
    use("@@/assets/")
    anvil, app, console, document, js, log, meta, native, packages, tools, window = (
        use.anvil,
        use.app,
        use.console,
        use.document,
        use.js,
        use.log,
        use.meta,
        use.native,
        use.packages,
        use.tools,
        use.window,
    )
    render = tools.render
    component = use("@@/component/")
    bootstrap = use("@/bootstrap/").bootstrap
    Toast = bootstrap.Toast

    class toast:
        def __init__(self):
            self._ = dict()

        def __call__(self, title: str, content: str, delay: int = 5000, style: str='primary'):
            element = component.div(
                "toast",
                role="alert",
                ariaLive="assertive",
                ariaAtomic="true",
                parent=self.container,
            )
            html = render(self.template, title=title, content=content, style=style)
            element.html(html)
            toast = js.new(Toast)(element, delay=delay)

            @element.on("hidden.bs.toast", once=True)
            def onhidden(event):
                """Cleans up."""
                ##log("Disposing")  ##
                event.stopPropagation()
                toast.dispose()
                element.remove()

            toast.show()

        @property
        def container(self):
            if not self._.get("container"):
                container = component.div(
                    "position-fixed.bottom-0.end-0.p-3.d-flex.flex-column.row-gap-3.z-3",
                    parent=app,
                )
                self._.update(container=container)
            return self._["container"]

        @property
        def template(self):
            if not self._.get("template"):
                template = use("assets/toast/toast.html", test=meta.test)
                self._.update(template=template)
            return self._["template"]

    toast = toast()

    return dict(toast=toast)
